internal/datasource/config/config_writer.go

```diff
@@ -7,108 +7,113 @@ package config
 
 import (
 	"context"
-	"encoding/json"
 	"fmt"
 	"log/slog"
 	"os"
 	"path"
 	"strings"
-	"time"
 
-	"github.com/nginx/agent/v3/internal/service/config"
+	"github.com/nginx/agent/v3/internal/config"
 
-	"github.com/google/uuid"
 	"github.com/nginx/agent/v3/api/grpc/instances"
 	"github.com/nginx/agent/v3/internal/client"
 )
 
 const (
-	cacheLocation        = "/var/lib/nginx-agent/config/%v/cache.json"
-	filePermissions      = 0o600
-	defaultClientTimeOut = 10 * time.Second
+	filePermissions = 0o600
 )
 
 //go:generate go run github.com/maxbrunsfeld/counterfeiter/v6@v6.7.0 -generate
 //counterfeiter:generate . ConfigWriterInterface
 
 type (
 	ConfigWriterInterface interface {
-		Write(ctx context.Context, filesURL string, tenantID uuid.UUID) (err error)
+		Write(ctx context.Context, filesURL, tenantID, instanceID string) (skippedFiles map[string]struct{}, err error)
 		Complete() (err error)
+		SetConfigClient(configClient client.ConfigClientInterface)
 	}
 
 	ConfigWriter struct {
 		configClient       client.ConfigClientInterface
-		previousFileCache  FileCache
-		currentFileCache   FileCache
-		cachePath          string
-		dataPlaneConfig    config.DataPlaneConfig
 		allowedDirectories []string
+		fileCache          FileCacheInterface
+		currentFileCache   CacheContent
 	}
-
-	// FileCache map of files with filepath as key
-	FileCache = map[string]*instances.File
 )
 
-func NewConfigWriter(configClient client.ConfigClientInterface,
-	allowedDirectories []string, instanceID string,
+func NewConfigWriter(agentConfig *config.Config, fileCache FileCacheInterface,
 ) *ConfigWriter {
-	cachePath := fmt.Sprintf(cacheLocation, instanceID)
-
-	if configClient == nil {
-		configClient = client.NewHTTPConfigClient(defaultClientTimeOut)
-	}
-
-	previousFileCache, err := readInstanceCache(cachePath)
-	if err != nil {
-		slog.Warn("Failed to Read cache %s ", cachePath, "err", err)
-		previousFileCache = nil
-	}
-
+	configClient := client.NewHTTPConfigClient(agentConfig.Client.Timeout)
```

Comment:
If `agentConfig` is nil, then this is a nil pointer dereference. 
```suggestion
        if agentConfig == nil {
            return nil
        }

	configClient := client.NewHTTPConfigClient(agentConfig.Client.Timeout)
```
Dunno if it would be a good idea to also return an error for the constructor in case the config is nil.

Reply:
Adding an error here probably make sense, since nothing will work really if you have no agentConfig

Reply:
Yeah I think it's a good convention to perform validations in the constructor and return errors if some of the params were invalid. It's very verbose with all the nil checks and you'll likely be checking the same config multiple times throughout the code base, but at least it's safe code.